[DCJ-457] Stairway can be built with a provided ThreadPoolTaskExecutor

okotsopoulos opened the pull request:
## Addresses
Partially addresses https://broadworkbench.atlassian.net/browse/DCJ-457

PRs will need to be merged in the following order:
1. This PR
2. https://github.com/DataBiosphere/terra-common-lib/pull/187
3. A forthcoming TDR PR to update its TCL version

## Summary of changes

This PR allows calling services to bring their own instrumented executor to serve as their Stairway's threadpool.

If unspecified in `StairwayBuilder`, a `DefaultThreadPoolTaskExecutor` will be constructed and used.

`DefaultThreadPoolTaskExecutor` is a public class: calling services may elect to instantiate it themselves (e.g. a Spring Boot service would register this as a Bean for automatic actuator instrumentation).

## Testing Strategy

Expanded unit tests.

Verified end-to-end behavior manually:
- Publishing Stairway and TCL to my local Maven repository
- Booting up TDR, with and without `maxParallelFlights` specified
- Observed that `stairwayExecutor` metrics were automatically flowing through TDR's Actuator endpoint
- Initiated a Stairway flight which succeeded.  Executor occupancy was reflected in actuator metrics.

Example where TDR configured its `maxParallelFlights` to be 120, has one actively running flight, and one completed flight:
```
(base) okotsopo@wm111-e35 jade-data-repo % curl http://localhost:8080/actuator/prometheus | grep 'stairwayExecutor'
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 87022  100 8executor_pool_core_threads{name="stairwayExecutor",} 120.0   0
7022    0     0  10.1M      0 --:--:-- --:--:-- --:--:-- 10.3M
executor_queue_remaining_tasks{name="stairwayExecutor",} 2.147483647E9
executor_completed_tasks_total{name="stairwayExecutor",} 1.0
executor_pool_size_threads{name="stairwayExecutor",} 2.0
executor_queued_tasks{name="stairwayExecutor",} 0.0
executor_pool_max_threads{name="stairwayExecutor",} 120.0
executor_active_threads{name="stairwayExecutor",} 1.0
```

okotsopoulos committed:
[DCJ-457] Stairway can be built with a provided ThreadPoolTaskExecutor

This allows calling services to bring their own instrumented executor to serve as their Stairway's threadpool.

If unspecified in StairwayBuilder, a DefaultThreadPoolTaskExecutor will be constructed and used.

DefaultThreadPoolTaskExecutor is a public class: calling services may elect to instantiate it themselves (e.g. a Spring Boot service would register this as a Bean for automatic actuator instrumentation).

okotsopoulos commented:
(Only marked as "ready for review" to see if that helps with checks not running… but it seems that GitHub is experiencing issues: https://www.githubstatus.com/incidents/9vwllhs2w1kj)

okotsopoulos committed:
Also assert that DefaultThreadPoolTaskExecutors are running

pshapiro4broad commented on stairway/src/main/java/bio/terra/stairway/impl/StairwayThreadPool.java:
```diff
@@ -3,34 +3,15 @@
 
-  int getQueuedFlights() {
-    return getQueue().size();
+  StairwayThreadPool(ThreadPoolTaskExecutor executor) {
```
I like the idea of moving the thread pool to a field rather than a base class. Is there a reason why you didn't continue to use `ThreadPoolExecutor` here? That way a user doesn't need to use a spring type to provide a custom thread pool. I believe the spring types are compatible with the java ones, so using `ThreadPoolExecutor` would still allow clients to use `ThreadPoolTaskExecutor` if they wanted to.

pshapiro4broad replied:
Reading on, it looks like this uses spring to track the number of active threads rather than doing it ourselves. I guess there's no harm in adding more spring dependencies here as we already use spring for all the java apps.

okotsopoulos replied:
Yes, that's the benefit that tipped the scale towards my usage of it over `ThreadPoolExecutor` (I liked that I was able to get rid of that custom tracking in this implementation as a result).  Tough to thread the needle sometimes between the diversity in tech stacks already present in our apps vs. what has become the prevailing standard (Spring Boot Java).

`ThreadPoolTaskExecutor` doesn't actually extend `ThreadPoolExecutor` either, so it would have been more involved (but not impossible) to support the non-Spring type.

nmalfroy replied:
Yeah, I had the same thought (and came to the same conclusion) for whatever that's worth :)

nmalfroy replied:
Another note: would it make sense to have a backwards compatibly constructor that takes an int and constructs the `ThreadPoolTaskExecutor`. That would make this PR not a breaking change?  I could see the value in explicitly making it a breaking change to force people to configure a threadpool but wanted to check

okotsopoulos replied:
That's a good question, but `StairwayThreadPool`s are not public, and we don't support their direct implementation by consumers.

I don't see this as a breaking change as far as our contracts with consumers are concerned, but rather an implementation detail, so I'm not inclined to maintain that old constructor.

nmalfroy replied:
OMG you're totally right.  Sorry, been looking at python too long and missed the absence of the `public` qualifier.

okotsopoulos replied:
No worries at all, I'm glad for the sanity check!

nmalfroy commented on DEVELOPMENT.md:
```diff
@@ -35,6 +35,36 @@ match the values in the sql file above:
+   # terra-workspace-manager/build.gradle
+   
+   repositories {
+     mavenLocal()
```
If testing locally like this is something common, would it make sense to have this controlled via a gradle property?  Something like
```suggestion
     if ("${properties["localTest"]}" == "true") {
        mavenLocal()
     }
     ...
```

And you could then use `-PlocalTest=true` to run that way.

Anyways, just a thought

okotsopoulos replied:
This was admittedly a copy-paste from the docs in terra-common-lib, I'll update to reflect the way TDR handles this (which is closer to what you've suggested).

rtitle commented on stairway/src/main/java/bio/terra/stairway/impl/StairwayThreadPool.java:
```diff
@@ -67,16 +48,4 @@ private void initializeFlightMdc(
     MdcUtils.removeStepContextFromMdc(flightContext);
   }
-
-  protected void beforeExecute(Thread t, Runnable r) {
```
I see these metrics were removed -- are these things we get "for free" with the switch to the Spring `ThreadPoolTaskExecutor`?

okotsopoulos replied:
Happily, yes!  And they're more valuable to us as a metric rather than a debug log.  They're available as the metric `executor_active_threads`.

```
(base) okotsopo@wm111-e35 jade-data-repo % curl http://localhost:8080/actuator/prometheus | grep 'stairwayExecutor' | grep 'executor_active_threads'
…
executor_active_threads{name="stairwayExecutor",} 1.0
```

nmalfroy commented on stairway/src/main/java/bio/terra/stairway/DefaultThreadPoolTaskExecutor.java:
```diff
@@ -0,0 +1,31 @@
+  }
+
+  private int getPoolSize(Integer maxParallelFlights) {
+    if (maxParallelFlights == null || maxParallelFlights <= 0) {
```
Just because it looks like you're switching using the syntax in StairwayImpl, this should probably be:
```suggestion
    if (Optional.ofNullable(maxParallelFlights).orElse(0) <= 0) {
```

But it's a whole character longer :)

okotsopoulos replied:
I'm actually going to walk back that syntax change, so I'll leave this line as it stands, thanks!

nmalfroy replied:
Absolutely.  Makes sense!

rtitle reviewed (approved):
This looks good to me -- I like the approach of decoupling `StairwayThreadPool` from the executor (instead of inheriting from it).

nmalfroy commented on stairway/src/main/java/bio/terra/stairway/impl/StairwayImpl.java:
```diff
@@ -105,6 +102,10 @@ public StairwayImpl(StairwayBuilder builder) throws StairwayExecutionException {
 
+    this.executor =
+        Optional.ofNullable(builder.getExecutor())
+            .orElse(new DefaultThreadPoolTaskExecutor(builder.getMaxParallelFlights()));
```
This is maybe minor but I think it's slightly better practice to do
```suggestion
            .orElseGet(() -> new DefaultThreadPoolTaskExecutor(builder.getMaxParallelFlights()));
```

By using a supplier, you don't actually create the `DefaultThreadPoolTaskExecutor` until you need it.  It's a constructor so maybe not a huge deal but I wanted to call it out since I've been bit by this before.

okotsopoulos replied:
I'm glad you raised this, I hadn't made that connection and agree that it should be changed.  This is getting into territory where the optional chaining is hindering readability for me, so I'll revisit in full.

nmalfroy commented on stairway/src/main/java/bio/terra/stairway/StairwayBuilder.java:
```diff
@@ -30,7 +32,7 @@ public class StairwayBuilder {
   /**
    * Determines the size of the thread pool used for running Stairway flights. Default is
-   * DEFAULT_MAX_PARALLEL_FLIGHTS (20 at this moment)
+   * established in {@link DefaultThreadPoolTaskExecutor} (20 at this moment).
```
Could just ponit to 
```suggestion
   * established in {@link DefaultThreadPoolTaskExecutor.DEFAULT_MAX_PARALLEL_FLIGHTS}
```

And I don't know that I'd put the value since that could get out of sync

okotsopoulos replied:
I had tried the link like this and it didn't work, but I figured out how to get it to work… much better :)

<img width="577" alt="Screenshot 2024-07-01 at 3 10 07 PM" src="https://github.com/DataBiosphere/stairway/assets/79769153/d16f611b-6bdb-40c2-874b-0f7a30e78ae5">
<img width="728" alt="Screenshot 2024-07-01 at 3 10 15 PM" src="https://github.com/DataBiosphere/stairway/assets/79769153/b3caf76c-8d7c-48ed-8530-86291d2a677c">

nmalfroy replied:
Ah good find!

nmalfroy reviewed (changes requested):
This looks great!  The reason I put in request changes is about some questions.  All minor but probably worthy of calling out

okotsopoulos committed:
Docstring references default maxParallelFlights directly

If this value were to change in the future, the javadoc rendered view would reflect the change.

okotsopoulos committed:
StairwayImpl constructor executor init follows existing convention

Mainly because my usage of Optional.ofNullable…orElse… constructs a new default executor every time, and throws it away if the builder has one set.  This is discouraged.

I decided to adhere with the existing convention for initializing instance variables from the StairwayBuilder object, which makes the constructor easier to follow.  If it's updated in the future with Optional handling, that would be acceptable as long as it's updated completely.

nmalfroy reviewed (approved):
You've answered all the questions I had :). Looks great!

okotsopoulos committed:
Update mavenLocal development docs

This matches TDR's current build file -- the useMavenLocal flag makes it a little easier to leverage locally published artifacts on demand.
